### PR TITLE
LUCENE-10153: Speed up BKDWriter using VarHandles.

### PR DESCRIPTION
**Pull request by jpountz**
See https://issues.apache.org/jira/browse/LUCENE-10153.

### CODE REVIEW

**Review by iverase — APPROVED** [review]
LGTM, really makes code easy to read.

We are always creating those object like:

```
 ArrayUtil.getUnsignedComparator(config.bytesPerDim)
```
I wonder as a follow up if we should move those object creation to the BKDConfig so we don't need to create them all over the place?

**Comment by iverase** [comment]
I am not worry about object creation but because we always call it with the same parameter (config.bytesPerDim), it seems to belong to config.

**Comment by jpountz** [comment]
I like that BKDConfig today only records data (bytesPerDim, etc.) rather than logic (how to compare byte[] arrays), so I have a bias towards keeping things as they are today.

File: lucene/core/src/java/org/apache/lucene/util/ArrayUtil.java [hunk 1]
```diff
@@ -694,4 +694,81 @@ public static int compareUnsigned4(byte[] a, int aOffset, byte[] b, int bOffset)
     return Integer.compareUnsigned(
         (int) BitUtil.VH_BE_INT.get(a, aOffset), (int) BitUtil.VH_BE_INT.get(b, bOffset));
   }
+
+  /**
+   * Return a comparator that computes the common prefix length across the next {@code numBytes} of
+   * the provided arrays.
+   */
+  public static ByteArrayComparator getPrefixLengthComparator(int numBytes) {
```
Comment (rmuir) [inline lucene/core/src/java/org/apache/lucene/util/ArrayUtil.java:702]: should all this fixed-length stuff really be here, or simply in BKD? Seems very specific to BKD.

Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/util/ArrayUtil.java]: This is a good point, I moved the new helpers to a pkg-private class.

Comment (rmuir) [reply lucene/core/src/java/org/apache/lucene/util/ArrayUtil.java]: Thanks! If we later need them for something outside of this package (e.g. a points query), then we can look at putting them in a more generic place.

Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/util/ArrayUtil.java]: +1

**Pull request merged**